docs/DifferentiableProgrammingImplementation.md

```diff
@@ -0,0 +1,1012 @@
+// Partially-applied to `@pb_foo` in `@vjp_foo`.
+struct foo_bb0_PB_src_0_wrt_0 {
+  let pb_sin: (Float) -> Float
+  let pb_cos: (Float) -> Float
+  let pb_mul: (Float) -> (Float, Float)
+}
+
+// VJP: replaces all function applications with VJP applications.
+sil @vjp_foo : $(Float) -> (Float, (Float) -> Float) {
+bb0(%x):
+  (%y1, %pb_sin) = apply @vjp_sin(%x)
+  (%y2, %pb_cos) = apply @vjp_cos(%x)
+  (%y3, %pb_mul) = apply @vjp_mul(%y1, %y2)
+  // Partially-apply to get a pullback.
+  %pb_struct = struct $foo_bb0_PB_src_0_wrt_0 (%pb_sin, %pb_cos, %pb_mul)
+  %pb = partial_apply @pb_foo(%pb_struct)
+  // Return tuple of original result and pullback.
+  %result = tuple (%y3, %pb)
+  return %result
+}
+
+// Pullback: apply pullbacks to adjoint values.
+sil @pb_foo : $(Float, foo_bb0_PB_src_0_wrt_0) -> (Float) {
+bb0(%dy3, %pb_struct):
+  // All "adjoint values" in the pullback are zero-initialized.
+  // %dx = 0, %dy1 = 0, %dy2 = 0
+  %pb_mul = struct_extract %pb_struct, #pb_mul
+  (%dy1, %dy2) += %pb_mul(%dy3)
+  %pb_cos = struct_extract %pb_struct, #pb_cos
+  (%dx) += %pb_cos(%dy2)
+  %pb_sin = struct_extract %pb_struct, #pb_sin
+  (%dx) += %pb_sin(%dy1)
+  return %dx
+}
+```
+
+## Derivative code generation details
+
+The sections above explain derivative code generation, focusing on the transformation rules for function applications (i.e. the `apply` instruction). However, SIL functions are often more complex than just “a single basic block of function applications”:
+
+- Original functions may involve control flow (conditionals, loops, `guard` and `switch` statements, etc) to express conditional operations. Such functions in SIL have multiple basic blocks and control-flow-related instructions.
+
+Multiple basic blocks and control flow basic block terminators (`cond_br`, `switch_enum`, etc) require special derivative code generation rules.
+
+> TODO: **Explain the control flow differentiation approach.** The approach is novel and highly worth documenting. [See “Control Flow Differentiation” slides](https://drive.google.com/corp/drive/folders/1y1FeJenXhoz_8xPw0ftVPngJUO_MdmRX) for an overview.
+
+- Original functions may have instructions other than `apply`. This includes the entire SIL [instruction set](https://github.com/apple/swift/blob/master/docs/SIL.rst#instruction-set):
+    - Memory-related operations: `alloc_stack`, `dealloc_stack`, `load`, `store`, etc.
+    - Aggregate operations: `struct`, `tuple`, `enum`.
+    - Projections: `struct_extract`, `tuple_extract`, `struct_element_addr`, `tuple_element_addr`.
+
+Many of these instructions have well-defined corresponding tangent instructions (for the differential) and adjoint instructions (for the pullback). It turns out that supporting transformations of a few common instructions is sufficient for many use cases (e.g. [deep learning APIs](https://github.com/tensorflow/swift-apis)). Here is a short table listing some of these instruction transformation rules:
```

Comment:
I'm kinda torn on this.  While the example of the swift-apis is useful, I think it might be better to remove the reference for now.

Reply:
Are you open to adding the reference back in when swift-apis becomes active again? We could also reference @ProfFan's SwiftFusion.

Reply:
I'd rather try to keep the links to the Swift projects.  If swift-apis becomes extremely active due to the community pushing it forward, perhaps that would be reasonable then.

Reply:
Could you clarify which choice you want? Keep as is or replace?

Reply:
I agree with @compnerd.  We should only use links to repos that are part of the Swift open source project (which tensorflow/swift-apis is not) or third-party projects that are actively maintained.

Reply:
@rxwei will hyperlinking to SwiftFusion be sufficient? If not, then we should leave it as is.

Reply:
I think that just removing it without a linked example is fine for the time being.

Reply:
```suggestion
Many of these instructions have well-defined corresponding tangent instructions (for the differential) and adjoint instructions (for the pullback). It turns out that supporting transformations of a few common instructions is sufficient for many use cases. Here is a short table listing some of these instruction transformation rules:
```